hredis/redigo/redigo.go

```diff
@@ -42,15 +44,61 @@ func WaitForAvailability(url string, d time.Duration, f WaitFunc) (bool, error)
 	}
 }
 
+// make this a var, so we can change this for testing.
+var redisDialURL = redis.DialURL
+
 // NewRedisPoolFromURL returns a new *redigo/redis.Pool configured for the supplied url
 // The url can include a password in the standard form and if so is used to AUTH against
 // the redis server
-func NewRedisPoolFromURL(url string) (*redis.Pool, error) {
+func NewRedisPoolFromURL(rawURL string, altPasses ...string) (*redis.Pool, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		findPassword sync.Once
+		password     string
+	)
+
+	// strip and save password
+	if pass, ok := u.User.Password(); ok {
+		password = pass
+	}
+
+	// DialURL will fail if wrong password is set. We want to create a successful connection
+	// with which we can try all of the possible passwords.
+	u.User = url.UserPassword("", "")
+	rawURL = u.String()
+
 	return &redis.Pool{
 		MaxIdle:     3,
 		IdleTimeout: 240 * time.Second,
 		Dial: func() (redis.Conn, error) {
-			return redis.DialURL(url)
+			c, err := redisDialURL(rawURL)
+			if err != nil {
+				return nil, err
+			}
+
+			findPassword.Do(func() {
```

Comment:
Ok how about this! Should what's here (trying all known passwords) happen every time Dial is called since the password the server wants could change while we're running? Based on my reading this will happen once for the lifetime of the pool/process.

Reply:
That probably makes more sense, generally speaking. If the connection has been "AUTH"'d successfully, and the password is changed _online_, I don't think the connection will be broken, but the next Dial won't get the right password. So, yeah, this makes sense. I'll update.

Reply:
Alright. The compromise here is to try first the last one we knew to be "good", falling back to the alternatives. When the "good" password is no longer valid, we nominate a new good one by trying them all.